src/java-interop/java-interop.targets

```diff
@@ -39,58 +24,64 @@
-      <_EnableMono>-DENABLE_MONO_INTEGRATION=ON</_EnableMono>
+    <PropertyGroup Condition=" $([MSBuild]::IsOSPlatform ('osx')) ">
+      <_CmakeOsxArch  Condition=" '$(NETCoreSdkRuntimeIdentifier)' == 'osx-x64'  ">x86_64</_CmakeOsxArch>
+      <_CmakeOsxArch  Condition=" '$(NETCoreSdkRuntimeIdentifier)' == 'osx-arm64'  ">arm64</_CmakeOsxArch>
     </PropertyGroup>
+
     <PropertyGroup>
-      <_JdkDirs>"-DJDK_INCLUDE_LIST=@(JdkIncludePath, ';')"</_JdkDirs>
-      <_Jni_c>"-DJNI_C_PATH=$(MSBuildThisFileDirectory)$(IntermediateOutputPath)jni.c"</_Jni_c>
-      <_ExtraArgs>$([MSBuild]::Escape('$(_JdkDirs) $(_Jni_c) $(_EnableMono) $(_MonoDirs) $(_MonoLib)'))</_ExtraArgs>
+      <_MonoNativePath>$(NuGetPackageRoot)microsoft.netcore.app.runtime.mono.$(NETCoreSdkRuntimeIdentifier)/$(DotNetRuntimePacksVersion)/runtimes/$(NETCoreSdkRuntimeIdentifier)/native/</_MonoNativePath>
+      <_MonoIncludePath>$(_MonoNativePath)include/mono-2.0</_MonoIncludePath>
+      <_DEnableMono>-DENABLE_MONO_INTEGRATION=ON</_DEnableMono>
+      <_DEnableOsxArchitectures Condition=" $([MSBuild]::IsOSPlatform ('osx')) ">"-DENABLE_OSX_ARCHITECTURES=$(_CmakeOsxArch)"</_DEnableOsxArchitectures>
+      <_DMonoDirs>"-DMONO_INCLUDE_LIST=$(_MonoIncludePath)"</_DMonoDirs>
+      <_DJdkDirs>"-DJDK_INCLUDE_LIST=@(JdkIncludePath, ';')"</_DJdkDirs>
+      <_DJni_c>"-DJNI_C_PATH=$(MSBuildThisFileDirectory)$(IntermediateOutputPath)jni.c"</_DJni_c>
+      <_MonoLinkFlags Condition=" $([MSBuild]::IsOSPlatform ('windows')) " >$(MSBuildThisFileDirectory)coreclr.lib</_MonoLinkFlags>
+      <_MonoLinkFlags Condition=" !$([MSBuild]::IsOSPlatform ('windows')) ">-L $(_MonoNativePath) -lcoreclr</_MonoLinkFlags>
+      <_DMonoLinkFlags>"-DMONO_LINK_FLAGS=$(_MonoLinkFlags)"</_DMonoLinkFlags>
+      <_ExtraArgs>$([MSBuild]::Escape('$(_DJdkDirs) $(_DJni_c) $(_DEnableMono) $(_DMonoDirs) $(_DMonoLinkFlags) $(_DEnableOsxArchitectures)'))</_ExtraArgs>
     </PropertyGroup>
   </Target>
 
+  <Target Name="_BuildJni_c"
+      Inputs="$(_JNIEnvGenPath)"
+      Outputs="$(IntermediateOutputPath)jni.c">
+    <MakeDir Directories="$(IntermediateOutputPath)" />
+    <Exec Command="$(_RunJNIEnvGen) $(IntermediateOutputPath)jni.g.cs $(IntermediateOutputPath)jni.c" />
+  </Target>
+
   <Target Name="_BuildLibs"
       Condition=" '$(NativeToolchainSupported)' == 'True' "
-      DependsOnTargets="GetNativeBuildCommands;_BuildJni_c;_GetCmakeOptions"
-      BeforeTargets="Build"
-      Inputs="@(_JavaInteropNativeLib);$(MSBuildThisFileFullPath);java-interop.csproj;@(ClInclude);@(ClCompile)"
-      Outputs="$(OutputPath)%(_JavaInteropNativeLib.Dir)$(_JavaInteropLibName)">
-    <MakeDir Directories="$(IntermediateOutputPath)%(_JavaInteropNativeLib.Dir)" />
+      DependsOnTargets="GetNativeBuildCommands;_BuildJni_c;_GetJavaInteropBuildInfo"
+      AfterTargets="ResolveReferences"
+      Inputs="CMakeLists.txt;$(MSBuildThisFileFullPath);java-interop.csproj;@(ClInclude);@(ClCompile)"
```

Comment:
Are the other files inputs like `coreclr.def`, `coreclr.lib`?

Reply:
I do not understand the question.

Conceptually, `coreclr.lib` could be added as an input, but (1) it should rarely change, and (2) if it *does* need to change, it's because you have a build failure anyway, e.g. because we added a call to a new `mono_whatever()` function that wasn't previously listed in `coreclr.def`/`coreclr.lib`.

It feels moot to add them, unless we add a Target which generates `coreclr.lib` from `coreclr.def`…

Reply:
I guess the question is if `coreclr.def/coreclr.lib` change, do you want `cmake` to run?

Reply:
If they change, it's because `cmake` errored out, in which case the `Outputs` won't exist, in which case updating `coreclr.*` won't matter, because re-building will re-run `cmake` anyway.

The only scenario this won't be the case is if you build successfully, then update the `.lib` to *remove* used symbols, and build again…

in which case, yeah, `coreclr.lib` *should* be an Input.  Doh!